From 6f5d4140a4e211cb651b51a041a43174f1aaa67a Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Tue, 11 Aug 2020 12:26:10 +0300
Subject: [PATCH] media: i2c: max9286: enable VS HS remap

This allows to remoap HS and VS pins using crossbar

Signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
---
 drivers/media/i2c/soc_camera/gmsl/max9286.c | 41 ++++++++++++++++++++++++-----
 1 file changed, 35 insertions(+), 6 deletions(-)

diff --git a/drivers/media/i2c/soc_camera/gmsl/max9286.c b/drivers/media/i2c/soc_camera/gmsl/max9286.c
index c37e972..8d9e72d 100644
--- a/drivers/media/i2c/soc_camera/gmsl/max9286.c
+++ b/drivers/media/i2c/soc_camera/gmsl/max9286.c
@@ -50,6 +50,8 @@ struct max9286_priv {
 	int			dbl;
 	int			dt;
 	int			hsgen;
+	int			vs;
+	int			hs;
 	u64			crossbar;
 	char			cb[16];
 	int			hts;
@@ -129,6 +131,14 @@ static int hsgen;
 module_param(hsgen, int, 0644);
 MODULE_PARM_DESC(hsgen, " Enable HS embedded generator (default: 0 - disabled)");
 
+static int hs = -1;
+module_param(hs, int, 0644);
+MODULE_PARM_DESC(hs, " HS input pin map to different DIN (default: -1 - disabled remap)");
+
+static int vs = -1;
+module_param(vs, int, 0644);
+MODULE_PARM_DESC(vs, " VS input pin map to different DIN (default: -1 - disabled remap)");
+
 static int pclk = 100;
 module_param(pclk, int, 0644);
 MODULE_PARM_DESC(pclk, " PCLK rate (default: 100MHz)");
@@ -464,15 +474,26 @@ static void max9286_gmsl_link_setup(struct i2c_client *client, int idx)
 
 			break;
 		}
-
+#if 0
+		/* these are defaults */
+		if (priv->ser_id == MAX96705_ID) {
+			reg8_write(client, 0x3f, 14);		/* HS (DIN14) */
+			reg8_write(client, 0x40, 15);		/* VS (DIN15) */
+		}
+		if (priv->ser_id == MAX96707_ID) {
+			reg8_write(client, 0x3f, 12);		/* HS (DIN12) */
+			reg8_write(client, 0x40, 13);		/* VS (DIN13) */
+		}
+#endif
+		/* HS/VS pins map */
+		if (priv->hs >= 0)
+			reg8_write(client, 0x3f, priv->hs);	/* HS (DIN_N) */
+		if (priv->vs >= 0)
+			reg8_write(client, 0x40, priv->vs);	/* VS (DIN_N) */
 		if (priv->hsgen) {
-			/* HS/VS pins map */
+			/* HS/DE are NC */
 			reg8_write(client, 0x3f, 0x10);			/* HS (NC) */
 			reg8_write(client, 0x41, 0x10);			/* DE (NC) */
-			if (priv->ser_id == MAX96705_ID)
-				reg8_write(client, 0x40, 15);		/* VS (DIN13) */
-			if (priv->ser_id == MAX96707_ID)
-				reg8_write(client, 0x40, 13);		/* VS (DIN13) */
 #if 0
 			/* following must come from imager */
 #define SENSOR_WIDTH	(1280*2)
@@ -768,6 +789,10 @@ static int max9286_parse_dt(struct i2c_client *client)
 		priv->dt = 3;
 	if (of_property_read_u32(np, "maxim,hsgen", &priv->hsgen))
 		priv->hsgen = 0;
+	if (of_property_read_u32(np, "maxim,hs", &priv->hs))
+		priv->hs = -1;
+	if (of_property_read_u32(np, "maxim,vs", &priv->vs))
+		priv->vs = -1;
 	if (of_property_read_u32(np, "maxim,pclk", &priv->pclk))
 		priv->pclk = pclk;
 	if (of_property_read_u32(np, "maxim,switchin", &priv->switchin))
@@ -802,6 +827,10 @@ static int max9286_parse_dt(struct i2c_client *client)
 		priv->dt = dt;
 	if (hsgen)
 		priv->hsgen = hsgen;
+	if (hs >= 0)
+		priv->hs = hs;
+	if (vs >= 0)
+		priv->vs = vs;
 	if (pclk != 100)
 		priv->pclk = pclk;
 	if (switchin)
-- 
2.7.4

